### PR TITLE
FIX: Handle Allocate for User-Defined StructInstanceMember

### PR DESCRIPTION
**Pull request by adit4443ya**
Towards 
https://github.com/lfortran/lfortran/issues/5655

## MRE 
----
```fortran
module p
    implicit none
    integer, parameter, private :: mxdim = 3
    type :: rp1d
        real(8), dimension(:), pointer :: f
    end type

    type :: sds
        integer, dimension(:), pointer :: dims
        type(rp1d), dimension(mxdim) :: scales
        integer :: f
    end type
end module

program struct_allocate
    use p
    implicit none
    type(sds) :: s
    !allocate(s%dims(2))
    ! allocate(s%scales(2)%f(2))
    ! s%scales(2)%f=[3,4]
    ! print *, 'Scale 0:', s%scales(2)%f(1)
    allocate(s%scales(1)%f(2))
    ! s%scales(1)%f=[1,2]
    ! print *, 'Scale 0:', s%scales(1)%f(1)
end program
```

### CODE REVIEW

**Comment by adit4443ya** [comment]
I have kept some lines commented in the test as of now just to see that what happens at CI
Eventually we will uncomment them in this PR.
Couple of things to consider 
- Is our ASR representation of 
`allocate(s%scales(1)%f(2))` is correct and if not then we should do refactor the ASR and then jump on LLVM backend 

Right now 
`allocate(s%scales(1)%f(2))`
seems like 
`allocate(s%scales%f(2))` in ASR there is no specification of the indices for s%scales in here hence i'm not sure if ASR representation is true

- The thins in Backend which I trying to fix in this PR, is to get the pointer of the element which is inturn the member of the structInstanceMember (Whose type is user defined) hence basically 

in Main branch 
`allocate(s%dims(2))` works well
But not 
`allocate(s%scales(1)%f(2))`

I have made changes so that allocation of these type can take place

- Next thing to consider is to uncomment the lines in test in Subsequent PR preferrably, as They provide some little edge cases which I can't explain in here.But Maybe needed to remove POT3D workaround https://github.com/gxyd/POT3D/commit/1aba062ac90c39e171f722c15b2bc6fafaaff251

**Review by kmr-srbh — CHANGES_REQUESTED** [review]
I think the changes are in the right direction. Apparently, it seems the incorrect representation of `allocate(s%scales(1)%f(2))` may be the issue we want to fix to get this working. Good catch!

**Comment by adit4443ya** [comment]
Here my approach is something like this 

For now hardcode the fortran_index=1 (for scales(1)) in the `s%scales(1)%f(2)`

So now we have `s%scales(1)%f(2)`

First inner recursive call (by this line `this->visit_expr(*x.m_v);`) for the `s%scales` will make `tmp` hold the address of the same 
then in the outer call we will get first get address of the `s%scales(1)` by 
`tmp = builder->CreateGEP(base_llvm_type, base_tmp, idx_vec);` where inx_vec would be ={0,0} //0-based indexing for llvm for fortran_index=1 e.g. scales(1)
Then create GEP for the `s%scales(1)%f`

But somhow the --fast flag is not working and current ci is failing with some Bus Error

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -2724,15 +2739,14 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
             current_der_type_name = dertype2parent[current_der_type_name];
         }
         int member_idx = name2memidx[current_der_type_name][member_name];
-
         llvm::Type *xtype = name2dertype[current_der_type_name];
         tmp = llvm_utils->create_gep2(xtype, tmp, member_idx);
         ASR::ttype_t* member_type = ASRUtils::type_get_past_pointer(
             ASRUtils::type_get_past_allocatable(member->m_type));
```
[the remark this answers is not in the document]
Comment (kmr-srbh) [reply src/libasr/codegen/asr_to_llvm.cpp]: Yes. I mean, could you please try getting `member_type` using `extract_type()`? Like `ASRUtils::extract_type(member->m_type)`.

[the remark this answers is not in the document]
Comment (adit4443ya) [reply src/libasr/codegen/asr_to_llvm.cpp]: Yes this diff works but not with `--fast` flag

```diff
-        ASR::ttype_t* member_type = ASRUtils::type_get_past_pointer(
-            ASRUtils::type_get_past_allocatable(member->m_type));
+        ASR::ttype_t *member_type = ASRUtils::extract_type(member->m_type);
```

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 2]
```diff
@@ -2716,6 +2720,25 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+            ASR::ttype_t* element_type = array_type->m_type;
+            
+            int fortran_index = 1; // TODO extract this from ASR for this case -> s%scales(2)%f 
+            llvm::Value* index = llvm::ConstantInt::get(context, llvm::APInt(32, fortran_index-1));
+            std::vector<llvm::Value*> idx_vec = {
+                llvm::ConstantInt::get(context, llvm::APInt(32, 0)),
+                index
+            };
+            llvm::Type* element_llvm_type = llvm_utils->get_type_from_ttype_t_util(element_type, module.get());
+            tmp = builder->CreateGEP(base_llvm_type, base_tmp, idx_vec);
```
Comment (Pranavchiku) [inline src/libasr/codegen/asr_to_llvm.cpp:2734]: Handle if differently for `FixedSizeArray` and `PointerToDataArray`

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (adit4443ya) [reply src/libasr/codegen/asr_to_llvm.cpp]: included this change in latest commit hence resolving this

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 2 above)
Comment (adit4443ya) [reply src/libasr/codegen/asr_to_llvm.cpp:2765]: I guess we would need that in the case of `s%scales(1)` while accessing 1st element of s%scales we would need to check if It's FixedSizearray or Descriptor one , as in `s%scales(1)%f` we are just accessing the `f` as member of the `s%scales(1)`
I have added check for the same in latest commit .
Let's hope CI passes now!!

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (adit4443ya) [reply src/libasr/codegen/asr_to_llvm.cpp]: CI failed hence removing that commit , as it wont work for aall cases, it will give asr_verify error

**Comment by gxyd** [comment]
So, I'm trying to first understand the ASR, and I think there is some fundamental problem in the created ASR itself, e.g.;
```f90
module p
    implicit none
    integer, parameter, private :: mxdim = 3
    type :: rp1d
        real(8), dimension(:), pointer :: f
    end type

    type :: sds
        type(rp1d), dimension(mxdim) :: scales
    end type
end module

program struct_allocate
    use p
    implicit none
    type(sds) :: s
    allocate(s%scales(2)%f(2))
end program
```

we've an which is:
```clj
(TranslationUnit
    (SymbolTable
        1
        {
            p:
                (Module
                    (SymbolTable
                        2
                        {
                            mxdim:
                                (Variable
                                    2
                                    mxdim
                                    []
                                    Local
                                    (IntegerConstant 3 (Integer 4) Decimal)
                                    (IntegerConstant 3 (Integer 4) Decimal)
                                    Parameter
                                    (Integer 4)
                                    ()
                                    Source
                                    Private
                                    Required
                                    .false.
                                    .false.
                                    .false.
                                ),
                            rp1d:
                                (Struct
                                    (SymbolTable
                                        3
                                        {
                                            f:
                                                (Variable
                                                    3
                                                    f
                                                    []
                                                    Local
                                                    ()
                                                    ()
                                                    Default
                                                    (Pointer
                                                        (Array
                                                            (Real 8)
                                                            [(()
                                                            ())]
                                                            DescriptorArray
                                                        )
                                                    )
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                    .false.
                                                    .false.
                                                )
                                        })
                                    rp1d
                                    []
                                    [f]
                                    Source
                                    Public
                                    .false.
                                    .false.
                                    []
                                    ()
                                    ()
                                ),
                            sds:
                                (Struct
                                    (SymbolTable
                                        4
                                        {
                                            scales:
                                                (Variable
                                                    4
                                                    scales
                                                    []
                                                    Local
                                                    ()
                                                    ()
                                                    Default
                                                    (Array
                                                        (StructType
                                                            2 rp1d
                                                        )
                                                        [((IntegerConstant 1 (Integer 4) Decimal)
                                                        (IntegerConstant 3 (Integer 4) Decimal))]
                                                        FixedSizeArray
                                                    )
                                                    ()
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                    .false.
                                                    .false.
                                                )
                                        })
                                    sds
                                    []
                                    [scales]
                                    Source
                                    Public
                                    .false.
                                    .false.
                                    []
                                    ()
                                    ()
                                )
                        })
                    p
                    []
                    .false.
                    .false.
                ),
            struct_allocate:
                (Program
                    (SymbolTable
                        5
                        {
                            1_rp1d_f:
                                (ExternalSymbol
                                    5
                                    1_rp1d_f
                                    3 f
                                    rp1d
                                    []
                                    f
                                    Public
                                ),
                            1_sds_scales:
                                (ExternalSymbol
                                    5
                                    1_sds_scales
                                    4 scales
                                    sds
                                    []
                                    scales
                                    Public
                                ),
                            rp1d:
                                (ExternalSymbol
                                    5
                                    rp1d
                                    2 rp1d
                                    p
                                    []
                                    rp1d
                                    Public
                                ),
                            s:
                                (Variable
                                    5
                                    s
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (StructType
                                        5 sds
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                    .false.
                                    .false.
                                ),
                            sds:
                                (ExternalSymbol
                                    5
                                    sds
                                    2 sds
                                    p
                                    []
                                    sds
                                    Public
                                )
                        })
                    struct_allocate
                    [p]
                    [(Allocate
                        [((StructInstanceMember
                            (StructInstanceMember
                                (Var 5 s)
                                5 1_sds_scales
                                (Array
                                    (StructType
                                        5 rp1d
                                    )
                                    [((IntegerConstant 1 (Integer 4) Decimal)
                                    (IntegerConstant 3 (Integer 4) Decimal))]
                                    FixedSizeArray
                                )
                                ()
                            )
                            5 1_rp1d_f
                            (Pointer
                                (Array
                                    (Real 8)
                                    [(()
                                    ())]
                                    DescriptorArray
                                )
                            )
                            ()
                        )
                        [((IntegerConstant 1 (Integer 4) Decimal)
                        (IntegerConstant 2 (Integer 4) Decimal))]
                        ()
                        ())]
                        ()
                        ()
                        ()
                    )]
                )
        })
    []
)
```

and if I make the below diff to the fortran program:
```diff
diff --git a/test.f90 b/test.f90
index 2a757ceba..b02f5d96e 100644
--- a/test.f90
+++ b/test.f90
@@ -14,5 +14,5 @@ program struct_allocate
     use p
     implicit none
     type(sds) :: s
-    allocate(s%scales(1)%f(2))
+    allocate(s%scales(2)%f(2))
 end program
```

I still get the same ASR, no change at all. So, there is some issue in the ASR itself here.

**Comment by adit4443ya** [comment]
> I still get the same ASR, no change at all. So, there is some issue in the ASR itself here.

Yes for that we would replace that with ArrayItem to get the indices of that scales in ASR

**Comment by gxyd** [comment]
The PR: https://github.com/lfortran/lfortran/pull/6510 is an attempt to fix the ASR first, let's see if the tests pass.

**Comment by Pranavchiku** [comment]
The PR is good, I approved it.

**Comment by gxyd** [comment]
We'll make use those changes in this PR, and make progress with this now.

**Comment by adit4443ya** [comment]
Thankyou @gxyd !!!!1
I will move this PR forward tonight

**Comment by Pranavchiku** [comment]
The added test works with latest main for me

```fortran
module mod_struct_allocate
    implicit none
    integer, parameter, private :: mxdim = 3
    type :: rp1d
        real(8), dimension(:), pointer :: f
    end type

    type :: sds
        type(rp1d), dimension(mxdim) :: scales
        integer, dimension(:), pointer ::dims
    end type
end module

program struct_allocate
    use mod_struct_allocate
    implicit none
    type(sds) :: s
    allocate(s%scales(1)%f(4))

    allocate(s%dims(2))
    s%dims=[10,11]
    print *, s%dims(1)

    allocate(s%scales(2)%f(3))
    s%scales(2)%f=[6,7,8]
    print *, s%scales(2)%f(3)

    allocate(s%scales(1)%f(2))
    s%scales(1)%f=[1,2]
    print *, s%scales(1)%f(1)
end program
```

```console
% lfortran a.f90 
10
8.00000000000000000e+00
1.00000000000000000e+00
% gfortran a.f90 && ./a.out
          10
   8.0000000000000000     
   1.0000000000000000     
```
